var/spack/repos/builtin/packages/mfem/package.py

```diff
@@ -366,6 +366,19 @@ def find_optional_library(name, prefix):
         cxxflags = spec.compiler_flags['cxxflags']
 
         if cxxflags:
+            # Add opt/debug flags if they are not present in global cxx flags
+            opt_flag_found = any(f in self.compiler.opt_flags for f in cxxflags)
+            debug_flag_found = any(f in self.compiler.debug_flags for f in cxxflags)
+
+            if '+debug' in spec:
+                if debug_flag_found:
+                    cxxflags.append('-g')
+                if opt_flag_found:
+                    cxxflags.append('-O0')
+            else:
+                if opt_flag_found:
+                    cxxflags.append('-O2')
```

Comment:
This kind of logic does not belong here. The variant `+debug`/`~debug` is intentionally independent of `cxxflags` if they are set by the spec. Overriding explicitly set values in `cxxflags` is not a good idea.

Note that the `debug` variant controls the internal mfem macro `MFEM_DEBUG` which enables a lot of run-time debug checks in the library. It affects the compilation options only if `cxxflags` is not set explicitly.

Reply:
@v-dobrev I think that what you've outlined doesn't satisfy the principle of least surprise to users. When people say `spack install foo+debug` they expect the package to be built with debug information. This PR ensures that is the case, and importantly it does NOT override what the user requested if they did request a debug flag.

I would agree with you if this PR overrode any flags found in the Spec, but it does not (minus typos). It only adds flags to bring the compiler flags into accordance with other aspects of the spec.

I expect to merge this in substantially its current form, once @white238 addresses the rest of the review.

Reply:
I agree, spack install mfem+debug should give us a a non-optimized build with symbols and the MFEM_DEBUG flag set to YES.

Reply:
If you want this kind of behavior with regard to `cxxflags`, this should be done by spack. You don't want such code in every package.

Reply:
It is true that Spack may help with some of these by providing helper functions or maybe some toggleable properties you can set on your package.  But in the state it was before, we could not debug MFEM on any of the very common LC development machine with the new set of intel or clang compilers, and that is not helpful for either of our teams.

Reply:
I don't follow, it seems you were already setting `cxxflags` in you debug builds, what stopped you from adding `-g` there?

Reply:
That was a typo and was corrected before it was merged.

Reply:
The cxxflags were set to a generic (for clang) "--gcc-toolchain=/path/to/newer/gcc".  This makes the compiler able to use the non-system, newer standard library.  This is because the default standard library on all (i think) TOSS3 machines is gcc 4.9.3.

This allows us to control which packages are debug at the package level instead of globally across all of our third party libraries.  Most of which we do not want to be debug.

Reply:
I was not referring to any changes in this PR. I was asking, before this was merged, why couldn't you just install mfem using something like: `spack install mfem+debug cxxflags="-g <other cxx flags>"`? Did that not work for some reason?

Reply:
I realized what you were saying in your previous comment and deleted my previous comment before this reply.. but to answer your question....

I am sure that would have worked but for the various teams I support, we go through a lot of effort to make our TPL builds as simple as possible for our development team.  There are many ways to do the same thing in Spack.  Also to refer back to @becker33 comment, the previous state was not what most Spack users would expect out of your package. It also doesn't change any behavior that was previously there.  It just adds flags that were possibly previously missing.

Reply:
For this example we build all of our TPLs by utilizing the command "spack dev-build -u host-config serac@develop+debug%clang@10.0.0".  This is also hidden in a script. If we specified all of the flags on every package that would be a lot of duplication of flags.

Reply:
Okay, I see that the issue was unexpected behavior: you expected `mfem+debug` will enable `-g` (even when `cxxflags` is explicitly given). On the other hand, in the mfem package we assumed that if `cxxflags` is specified then we should use it as given and not make any modifications (except things that are absolutely necessary like adding the c++11 flag).

I think that the right way to ensure this convention (`+debug` variant implies `-g` in the compiler options) is followed is to do it in spack. Either by extending the `cxxflags` given to the package, or even more forcefully by injecting `-g` (or equivalent) through the compiler wrapper. Otherwise, this same logic will have to be implemented by every package individually.

Reply:
I agree you don't want this logic to spread everywhere.  I would be interested to know of how a CMakePackage handles this since CMake has baked in build types that are separate from the base compiler flags.  I will sometime play around with one to see how it reacts but it's dinner time.